csrc/scheduler/registry_utils.cpp

```diff
@@ -572,59 +572,96 @@ bool SchedulerTopologyChecker::hasNonNormalizePostReductionBCast(
+          if (validated_resolved_tvs.count(forward_running_consumer)) {
+            continue;
+          }
 
-        // If consumer was already validated it doesn't need to be checked
-        if (validated_resolved_tvs.count(forward_running_consumer)) {
-          continue;
-        }
+          auto forward_pairwise_root_map = PairwiseRootDomainMap(
+              forward_running_producer, forward_running_consumer);
+          auto forward_p2c_root_map =
+              forward_pairwise_root_map.mapProducerToConsumer();
 
-        auto forward_pairwise_root_map = PairwiseRootDomainMap(
-            forward_running_producer, forward_running_consumer);
-        auto forward_p2c_root_map =
-            forward_pairwise_root_map.mapProducerToConsumer();
-
-        // These are the ids we will have to resolve. As we resolve them we'll
-        // remove them from this vector. If this vector ends up empty, then
-        // we've resolved everything we need to. This is a pair so as we
-        // traverse we can map the id through the traversal. The first entry
-        // in the pair will be the original id so we can reset it if it's not
-        // resolved before the next traversal. The second ID will be
-        // propagated as we map the IDs through the backward traversal.
-        std::vector<std::pair<IterDomain*, IterDomain*>> ids_to_resolve;
-
-        // Check if any TensorViews have a resolved broadcast
-        for (auto entry : forward_p2c_root_map) {
-          auto p_id = entry.first;
-          auto c_id = entry.second;
-          if (p_id->isBroadcast() && !c_id->isBroadcast()) {
-            ids_to_resolve.emplace_back(c_id, c_id);
+          // Check if any TensorViews have a resolved broadcast
+          for (auto entry : forward_p2c_root_map) {
+            auto p_id = entry.first;
+            auto c_id = entry.second;
+            if (p_id->isBroadcast() && !c_id->isBroadcast()) {
+              ids_to_resolve.emplace_back(c_id, c_id);
+            }
+          }
+          if (ids_to_resolve.empty()) {
+            continue;
           }
         }
 
-        if (ids_to_resolve.empty()) {
-          continue;
+        // At this point, we have some ids to resovle.
+        // Check if we can move forward further along the use chain.
+        // This helps gathering more usage info to resolve this id to avoid
+        // unnecessary segmentation when the reduction input iteration domain is
+        // used after this tensor. See PostReductionBroadcastCheck test and
+        // issue-2146. We can move forward when:
+        // (1) The usage chain is not empty.
+        // (2) The consumer is not a broadcastOp, as broadcastOp introduces a
+        //     new broadcast axis and require a new check, see
+        //     FusionReshapePersistentShmoo.
+        // (3) At least one id is mapped, otherwise can't do backward traversal.
+        if (!forward_tv_dep_chain.empty()) {
+          auto tmp_producer = forward_running_consumer;
```

Comment:
When this line is visited with non-empty `ids_to_resolve` as part of the while loop of line 586, shouldn't it use `forward_running_consumer` and `forward_running_producer` of lines 587 and 588?

Reply:
Here we are trying to moveforward along the use chain. For example when we have
 `T0 (forward_running_producer)-> T1 (forward_running_consumer)  -> T2`
At this point, we checked `T0->T1` and want to check if we may moveforward to `T1->T2`, so `T1` becomes the `tmp_producer ` and `T2` becomes the `tmp_consumer`

Reply:
Maybe I'm reading the code incorrectly, but what happens if there's also another consumer following `T2`, say `T3`? The pair of `T1` and `T2` is checked here and then the control goes back to the top of the loop body of the while loop starting at line 586, right? Even though there's `T3` left, since `forward_tv_dep_chain` is empty since there's `pop_front`, `T3` is not going to be visited, right?

Reply:
we didn't pop front when trying to move forward, it's just a copy of the front element and see wheter we can do it.
```
          auto tmp_producer = forward_running_consumer; (which is T1)
          auto tmp_consumer = forward_tv_dep_chain.front(); (which is T2)
          T2 is NOT popped.
```
In your case, when go back to the top of the loop body of the while loop, 
```
        forward_running_producer = forward_running_consumer; (which is T1)
        forward_running_consumer = forward_tv_dep_chain.front(); (which is T2)
        T2 is popped.
```

Reply:
Oh, I see. It seems this while loop essentially consists of two separate parts. Wouldn't it make more sense to have two while loops, one is to find a resolution, i.e., line 598 to line 628, and another is to move forward further.

Reply:
That's what I did at the beginning of this implementation. Let me roll back to the original version.

Reply:
That's not what I meant. The previous version has a nested while loop with the same loop condition. What I'm suggesting is to have two separate while loops.

Reply:
oh, got you. let me check how to do that.

Reply:
Revised. We need 3 while loops. The current logic is:
```
// check this dep chain
while (!forward_tv_dep_chain.empty()) {
  (1) step-1, Find the id to resolve
  while{}
  (2) step-2, Find the location we need to start the backward traversal resolve
  while{}
  (3) step-3, do backward traversal resolve
  (4) step-4, move to beg of the outer while loop, continue check of the dep chain
}
```